feature(web_services): allows calling WS functions with associative array

mrclay opened the pull request:
(this was replaced by #9418)

Plugins can expose their methods in a way that invokes callback functions with a single argument that contains an associative array of parameters, rather than each parameter being its own argument.

Deprecates verify_parameters() and serialise_parameters() that are no longer in use.

Updates unit tests to include new features, as well as fixes some code issues to match test expectations

Allows all callables as method callback.

(this rebases #8615 and makes it mostly BC with 2.0)
- [ ] this fixes #9411. Can we do this in 2.x?
- [x] verify "required" has the same default as in 2.x
- [x] duplicate 2.x casting behavior
- [ ] re-implement these changes in simplest way

hypeJunction committed:
feature(web_services): allows calling WS functions with associative array

Plugins can expose their methods in a way that invokes callback functions
with a single argument that contains an associative array of parameters,
rather than each parameter being its own argument.

Deprecates verify_parameters() and serialise_parameters() that are no longer
in use.

Updates unit tests to include new features, as well as fixes some code issues
to match test expectations.

Removes eval() magic and serialization from web services to allow for a wider
range of callables as method callback.

Improves handling and validation of parameters.

beck24 commented:
Awesome, this is a feature I'll be using right away.
LGTM

hypeJunction commented:
I don't see what changed from the original PR :)

juho-jaakkola commented on mod/web_services/tests/ElggCoreWebServicesApiTest.php:
```diff
@@ -169,51 +174,52 @@ public function testExecuteMethodWrongMethod() {
+//			$this->assertIsA($e, 'APIException');
+//			$this->assertIdentical($e->getMessage(), sprintf(elgg_echo('APIException:InvalidParameter'), 'param1', 'test'));
+//		}
+//	}
```
Why has this been commented?

juho-jaakkola commented on docs/guides/web-services.rst:
```diff
@@ -113,26 +118,63 @@ You can use additional fields to describe your parameter, e.g. ``description``.
+		'GET',
+		false,
+		false,
+		true // $assoc makes the callback receive an associative array
```
I haven't looked into the details yet, but wouldn't it be possible to check whether a parameter is an array or not?

mrclay replied:
Individual parameters can be arrays, too, so we can't tell by reflection. We could better do this by having plugins just provide an instance of an interface (with isAssoc(), etc.) for each method, but the goal here is to nudge this along instead of getting bogged down.

mrclay commented:
I'm going to move some of these changes to a 2.0 PR.

mrclay commented:
Please see #9411. If we decide it can be fixed in 2.x we can merge this as is.

mrclay committed:
chore(web_services): fix various BC issues

Casting follows 2.0 behavior and adds more thorough tests.
Throw InvalidParameterException if non-array passed in.
Empty arrays are valid as array input.
An explicit "default" key must be given as a default to be used.
A falsy default value should not be considered missing.
Re-add testVerifyParametersTypeNotSet.

mrclay commented:
For more BC with 2.0 I've added https://github.com/Elgg/Elgg/commit/b07e6453c3869a9b90f5a2dcb2e08451ae732a18. So basically this PR mostly fixes #9411 and allows all callables. Reforming how string values are cast would have a big BC impact. If that's something urgent the desired behavior should be documented in a separate issue.

mrclay commented:
The feature/fix this PR provides churns the code a bunch for BC concerns and I'd rather put a cleaner PR together to avoid accidentally modifying behavior in 2.x.

hypeJunction commented:
We are back to the question if it's worth the effort. You might as well invest that time into a complete rewrite.

mrclay commented:
I can give it another few min, worst case I'll just reopen this.

beck24 commented:
wasn't one of the arguments for separating the core plugins that we don't have to be too concerned about BC within plugins?  We can release a new major semver release of webservices, people who really require the old behavior can just use the old one.  Maybe this should/could be a trial run with that.

mrclay commented:
@beck24 yeah we can do that. #9419
